### PR TITLE
Show “invite by SMS” offer for matching non-Signal contacts when sear…

### PR DESCRIPTION
**Pull request by charlesmchen**
…ching in “new 1:1: conversation” view.

PTAL @michaelkirk 

img_0281

### CODE REVIEW

**Comment by charlesmchen** [comment]
@moxie0 This was a neat idea from @michaelkirk.  You might want to do something similar on Android.

**Comment by charlesmchen** [comment]
Also including: Slightly reduce the non-contact cell heights in “new 1:1 conversation” view.

img_0032

File: Signal/src/ViewControllers/ContactsViewHelper.m [hunk 1]
```diff
@@ -199,6 +207,76 @@ - (BOOL)doesSignalAccount:(SignalAccount *)signalAccount matchSearchTerms:(NSArr
...
+
+    return NO;
+}
+
+- (BOOL)doesContact:(Contact *)contact matchSearchTerms:(NSArray<NSString *> *)searchTerms
+{
+    OWSAssert(contact);
+    OWSAssert(searchTerms.count > 0);
+
+    for (NSString *searchTerm in searchTerms) {
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ContactsViewHelper.m]: why does this code feel so familiar?

File: Signal/src/ViewControllers/ContactsViewHelper.m [hunk 2]
```diff
@@ -199,6 +207,76 @@ - (BOOL)doesSignalAccount:(SignalAccount *)signalAccount matchSearchTerms:(NSArr
...
+        return [NSArray new];
+    }
+
+    return [self.nonSignalContacts filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(Contact *contact,
+                                                                   NSDictionary<NSString *, id> *_Nullable bindings) {
+        return [self doesContact:contact matchSearchTerms:searchTerms];
+    }]];
+}
+
+- (nullable NSArray<Contact *> *)nonSignalContacts
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ContactsViewHelper.m]: this might be an empty array, but it won't be nil, right?

**Comment by charlesmchen** [comment]
> when you hit cancel the searchfield is empty, but the list is still filtered

I can't repro this issue. I've tested iOS 9 and 10.  Can you give me any more details?

> I'm sure this was a pre-existing issue, but SMS view is illegible . We need to "apply default appearance" like we do when messaging someone via the "Invite Friends to Signal" flow (then revert to signal appearance upon completion). You could also try using the InviteFlow.sendSmsTo(phoneNumbers:[String]) which should take care of this for you.

I've filed a card.

File: Signal/src/ViewControllers/ContactsViewHelper.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/ContactsViewHelper.m]: We do something similar when searching for `SignalContacts`.  There's just enough differences that it doesn't make sense to de-duplicate, IMO.

File: Signal/src/ViewControllers/ContactsViewHelper.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/ContactsViewHelper.m]: Well, this value is cached and we clear the cache when contacts change by setting the cached value to nil.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

**Comment by charlesmchen** [comment]
> when you hit cancel the searchfield is empty, but the list is still filtered

I'm going to merge, but can you please re-test this at some point?  I can't repro.  I'll make sure it's in the test plan too.

**Pull request merged**